apps/aecore/src/aec_target.erl

```diff
@@ -28,40 +28,77 @@ determine_delta_header_height(Header) ->
+%%                  = HIGHEST_TARGET * DesiredRate / EstimatedCapacity
+%%
+%% We can estimate the network capacity used to mine a given block `i` as
+%%
+%%    EstimatedCapacity[i] = Difficulty[i] / MiningTime[i]
+%%    MiningTime[i]        = Time[i + 1] - Time[i]
+%%
+%% The estimated capacity across all `N` blocks is then the weighted (by time)
+%% average of the estimated capacities for each block.
+%%
+%%    EstimatedCapacity = Sum(EstimatedCapacity[i] * MiningTime[i]) / TotalTime
+%%                      = Sum(Difficulty[i]) / TotalTime
+%%                      = Sum(HIGHEST_TARGET / Target[i]) / TotalTime
+%%
+%% Now, the problem is that we can't do any floating point arithmetic (to
+%% ensure the calculation can be verified by other nodes), so we pick a
+%% reasonably big integer K (= HIGHEST_TARGET * 2^32) and compute
+%%
+%%    EstimatedCapacity ≈ Sum(K * HIGHEST_TARGET div Target[i]) / TotalTime / K
+%%
+%% Then
+%%
+%%    NewTarget = HIGHEST_TARGET * DesiredRate / EstimatedCapacity
+%%              ≈ HIGHEST_TARGET * DesiredRate * TotalTime * K / Sum(K * HIGHEST_TARGET div Target[i])
+%%              ≈ DesiredRate * TotalTime * K / Sum(K div Target[i])
+%%              ≈ TotalTime * K div (DesiredTimeBetweenBlocks * Sum(K div Target[i]))
+%%
+recalculate(Top, PrevHeaders) ->
+    K                        = aec_pow:scientific_to_integer(?HIGHEST_TARGET_SCI) * (1 bsl 32),
+    SumKDivTargets           = lists:sum([ K div aec_pow:scientific_to_integer(aec_headers:target(Hd))
+                                           || Hd <- PrevHeaders ]),
+    DesiredTimeBetweenBlocks = aec_governance:expected_block_mine_rate(),
+    Last                     = lists:last(PrevHeaders),
+    TotalTime                = mining_time_between(Last, Top),
+    NewTargetInt             = TotalTime * K div (DesiredTimeBetweenBlocks * SumKDivTargets),
+    min(?HIGHEST_TARGET_SCI, aec_pow:integer_to_scientific(NewTargetInt)).
 
--spec verify(header(), header(), header()) -> ok | {error, target_too_high}.
-verify(Header, PrevHeader, InitialHeader) ->
-    HeaderTarget = aec_headers:target(Header),
-    PrevTarget   = aec_headers:target(PrevHeader),
-
-    %% Block candidate target was computed using the target of the previous block
-    %% we have to do the same thing here.
-    ExpectedTarget = recalculate(aec_headers:set_target(Header, PrevTarget),
-                                 InitialHeader),
-
-    %% Currently we don't verify if target was too low (difficulty was too high),
-    %% so that someone might have tried to send a block
-    %% with too high difficulty (too low target).
-    %% TODO: Investigate whether we need to avoid "too difficult" blocks
-    case HeaderTarget =< ExpectedTarget of
+-spec verify(header(), nonempty_list(header())) -> ok | {error, wrong_target}.
+verify(Top, PrevHeaders) ->
+    HeaderTarget = aec_headers:target(Top),
+    ExpectedTarget = recalculate(Top, PrevHeaders),
+    case HeaderTarget == ExpectedTarget of
```

Comment:
Is this correct? Don't we accept targets that is more difficult? If another miner chooses to do so?

Reply:
Answering myself: The consensus is that this is correct.

Reply:
I would actually prefer checking exact target - that I had tracked as https://www.pivotaltracker.com/story/show/153258322